File: leetcode/2024_7_july/python/12_maximum_score_from_removing_substring.py
# CODE

class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        res = 0
        bade = 'ab' if x > y else 'ba'
        chote = 'ba' if bade == 'ab' else 'ab'

        # remove substring function
        def remove(s, t):
            stk = []
            for ch in s:
                if (ch == t[1] and stk and stk[-1] == t[0]):
                    stk.pop()
                else:
                    stk.append(ch)
            return ''.join(stk)

        # first pass
        remove_pair_str = remove(s, bade)
        remove_pair_cnt = (len(s) - len(remove_pair_str)) // 2
        res += (remove_pair_cnt * max(x, y))

        # second pass
        s = remove(remove_pair_str, chote)
        remove_pair_cnt = (len(remove_pair_str) - len(s)) // 2
        res += (remove_pair_cnt * min(x, y))

        return res

File: leetcode/2024_7_july/c++/12_maximum_score_from_removing_substring.cpp
// TIME COMPLEXITY
// O(n)

// SPACE COMPLEXITY
// O(n)

// CODE

class Solution {
public:
    int maximumGain(string s, int x, int y) {
        int flag = 0;
        if(x > y) flag = 1;
        int res = 0;

        stack<char> st1;
        stack<char> st2;

        for(auto c : s){
            if(!st1.empty()){
                if(flag == 1){
                    if(st1.top() == 'a' && c == 'b'){
                        res += x;
                        st1.pop();
                    }
                    else{
                        st1.push(c);
                    }
                }
                else{
                    if(st1.top() == 'b' && c == 'a'){
                        res += y;
                        st1.pop();
                    }
                    else{
                        st1.push(c);
                    }
                }
            }
            else{
                st1.push(c);
            }
        }

        string remaining;
        while (!st1.empty()) {
            remaining += st1.top();
            st1.pop();
        }
        reverse(remaining.begin(), remaining.end());

        for(auto c : remaining){
            if(!st2.empty()){
                if(!flag){
                    if(st2.top() == 'a' && c == 'b'){
                        res += x;
                        st2.pop();
                    }
                    else{
                        st2.push(c);
                    }
                }
                else{
                    if(st2.top() == 'b' && c == 'a'){
                        res += y;
                        st2.pop();
                    }
                    else{
                        st2.push(c);
                    }
                }
            }
            else{
                st2.push(c);
            }
        }

        return res;


    }
};